File: samples/vdsp_op/batch_crop_resize/batch_crop_resize.py
import vaststreamx as vsx
import numpy as np
import cv2
import argparse
import ast

attr = vsx.AttrKey


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/dog.jpg",
        help="input image",
    )
    parser.add_argument(
        "--output_size", help="resize output size [w,h]", default="[512, 512]"
    )
    parser.add_argument(
        "--output_file1",
        default="batch_crop_resize_result1.jpg",
        help="output image 1",
    )
    parser.add_argument(
        "--output_file2",
        default="batch_crop_resize_result2.jpg",
        help="output image 2 ",
    )
    parser.add_argument(
        "--crop_rect1",
        default="[50,70,131,230]",
        help="crop rect [x,y,w,h]",
    )
    parser.add_argument(
        "--crop_rect2",
        default="[60,90,150,211]",
        help="crop rect [x,y,w,h]",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    vsx.set_device(args.device_id)
    cv_image = cv2.imread(args.input_file)
    assert cv_image is not None, f"Failed to read: {args.input_file}"
    vsx_image = vsx.create_image(
        cv_image,
        vsx.ImageFormat.BGR_INTERLEAVE,
        cv_image.shape[1],
        cv_image.shape[0],
        args.device_id,
    )
    # Build operator
    op = vsx.BuildInOperator(vsx.OpType.SINGLE_OP_BATCH_CROP_RESIZE)

    output_size = ast.literal_eval(args.output_size)
    crop_rect1 = ast.literal_eval(args.crop_rect1)
    crop_rect2 = ast.literal_eval(args.crop_rect2)

    rect_num = 2
    op.set_attribute(
        {
            attr.IIMAGE_FORMAT: vsx.ImageType.BGR888,
            attr.IIMAGE_WIDTH: vsx_image.width,
            attr.IIMAGE_HEIGHT: vsx_image.height,
            attr.IIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.IIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.OIMAGE_WIDTH: output_size[0],
            attr.OIMAGE_HEIGHT: output_size[1],
            attr.OIMAGE_WIDTH_PITCH: output_size[0],
            attr.OIMAGE_HEIGHT_PITCH: output_size[1],
            attr.RESIZE_TYPE: vsx.ImageResizeType.BILINEAR_CV,
            attr.CROP_NUM: rect_num,
            attr.CROP_X: [crop_rect1[0], crop_rect2[0]],
            attr.CROP_Y: [crop_rect1[1], crop_rect2[1]],
            attr.CROP_WIDTH: [crop_rect1[2], crop_rect2[2]],
            attr.CROP_HEIGHT: [crop_rect1[3], crop_rect2[3]],
        }
    )
    print(f"get attribute(attr.CROP_NUM):{op.get_attribute(attr.CROP_NUM)}")
    # run
    output_list = op.execute(
        [vsx_image],
        [(vsx.ImageFormat.BGR_INTERLEAVE, output_size[0], output_size[1])] * rect_num,
    )

    out_images = [vsx.as_numpy(output).squeeze() for output in output_list]
    cv2.imwrite(args.output_file1, out_images[0])
    cv2.imwrite(args.output_file2, out_images[1])
    print(f"save result to {args.output_file1}, {args.output_file2}")

File: samples/vdsp_op/batch_crop_resize/batch_crop_resize.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <chrono>
#include <thread>

#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/dog.jpg");
  args.add<std::string>("output_file1", '\0', "output image", false,
                        "batch_crop_resize_result1.jpg");
  args.add<std::string>("output_file2", '\0', "output image", false,
                        "batch_crop_resize_result2.jpg");
  args.add<std::string>("output_size", '\0', "output size [w,h]", false,
                        "[512,512]");
  args.add<std::string>("crop_rect1", '\0', "crop rect1 [x,y,w,h]", false,
                        "[50,70,131,230]");
  args.add<std::string>("crop_rect2", '\0', "crop rect1 [x,y,w,h]", false,
                        "[60,90,150,211]");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  uint32_t device_id = args.get<uint32_t>("device_id");
  auto input_file = args.get<std::string>("input_file");
  std::vector<uint32_t> crop_rect1 =
      vsx::ParseVecUint(args.get<std::string>("crop_rect1"));
  std::vector<uint32_t> crop_rect2 =
      vsx::ParseVecUint(args.get<std::string>("crop_rect2"));
  std::vector<uint32_t> output_size =
      vsx::ParseVecUint(args.get<std::string>("output_size"));

  vsx::SetDevice(device_id);

  vsx::Image image;
  if (vsx::MakeVsxImage(input_file, image, vsx::BGR_INTERLEAVE) != 0) {
    std::cout << "Read input image failed: " << input_file << std::endl;
  }

  int crop1_x = crop_rect1[0];
  int crop1_y = crop_rect1[1];
  int crop1_w = crop_rect1[2];
  int crop1_h = crop_rect1[3];

  int crop2_x = crop_rect2[0];
  int crop2_y = crop_rect2[1];
  int crop2_w = crop_rect2[2];
  int crop2_h = crop_rect2[3];

  int oimage_width = output_size[0];
  int oimage_height = output_size[1];

  int rects = 2;
  auto op = vsx::BuildInOperator(
      vsx::BuildInOperatorType::kSINGLE_OP_BATCH_CROP_RESIZE);

  // [4] set attr

  op.SetAttribute<vsx::AttrKey::kIimageFormat>(
      vsx::BuildInOperatorAttrImageType::kBGR888);
  op.SetAttribute<vsx::AttrKey::kIimageWidth>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeight>(image.Height());
  op.SetAttribute<vsx::AttrKey::kIimageWidthPitch>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeightPitch>(image.Height());
  op.SetAttribute<vsx::AttrKey::kOimageWidth>(oimage_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeight>(oimage_height);
  op.SetAttribute<vsx::AttrKey::kOimageWidthPitch>(oimage_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeightPitch>(oimage_height);
  op.SetAttribute<vsx::AttrKey::kResizeType>(
      vsx::ImageResizeType::kRESIZE_TYPE_BILINEAR_CV);

  op.SetAttribute<vsx::AttrKey::kCropNum>(rects);

  op.SetAttribute<vsx::AttrKey::kCropX>(crop1_x, 0);
  op.SetAttribute<vsx::AttrKey::kCropY>(crop1_y, 0);
  op.SetAttribute<vsx::AttrKey::kCropWidth>(crop1_w, 0);
  op.SetAttribute<vsx::AttrKey::kCropHeight>(crop1_h, 0);

  op.SetAttribute<vsx::AttrKey::kCropX>(crop2_x, 1);
  op.SetAttribute<vsx::AttrKey::kCropY>(crop2_y, 1);
  op.SetAttribute<vsx::AttrKey::kCropWidth>(crop2_w, 1);
  op.SetAttribute<vsx::AttrKey::kCropHeight>(crop2_h, 1);

  // execute

  // copy input to device memory
  auto input_vacc = vsx::Image(image.Format(), image.Width(), image.Height(),
                               vsx::Context::VACC(device_id));
  input_vacc.CopyFrom(image);

  // make device memory image for output
  auto output1_vacc = vsx::Image(image.Format(), oimage_width, oimage_height,
                                 vsx::Context::VACC(device_id));
  auto output2_vacc = vsx::Image(image.Format(), oimage_width, oimage_height,
                                 vsx::Context::VACC(device_id));
  std::vector<vsx::Image> outputs{output1_vacc, output2_vacc};
  // buildin operator execute
  op.Execute({input_vacc}, outputs);
  // copy result to cpu memory
  vsx::Image output1_cpu = vsx::Image(
      output1_vacc.Format(), output1_vacc.Width(), output1_vacc.Height());
  output1_cpu.CopyFrom(output1_vacc);

  vsx::Image output2_cpu = vsx::Image(
      output2_vacc.Format(), output2_vacc.Width(), output2_vacc.Height());
  output2_cpu.CopyFrom(output2_vacc);

  cv::Mat out1_mat, out2_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output1_cpu, out1_mat);
  vsx::ConvertVsxImageToCvMatBgrPacked(output2_cpu, out2_mat);
  cv::imwrite(args.get<std::string>("output_file1"), out1_mat);
  cv::imwrite(args.get<std::string>("output_file2"), out2_mat);
  std::cout << "save result to : " << args.get<std::string>("output_file1")
            << ", " << args.get<std::string>("output_file2") << std::endl;
  return 0;
}
